src/inmanta/ast/variables.py

```diff
@@ -101,111 +101,87 @@ def __repr__(self) -> str:
 
-    def target_value(self) -> T:
+    def write_target(self, variable: ResultVariable[object]) -> None:
         """
-        Returns the target value based on self.variable
+        Writes the target variable based on the complete variable's value.
         """
-        raise NotImplementedError()
+        self.target.set_value(self.target_value(variable), self.location)
 
-    def resume(self, requires: Dict[object, ResultVariable], resolver: Resolver, queue_scheduler: QueueScheduler) -> None:
+    def target_value(self, variable: ResultVariable[object]) -> T:
         """
-        Instance is ready to execute, do it and see if the variable is already present
+        Returns the target value based on the complete variable's value.
         """
-        if not self.variable:
-            # this is a first time we are called, variable is not cached yet
-            self.variable = self.fetch_variable(requires, resolver, queue_scheduler)
+        return variable.get_value()
 
+    def resume(
+        self,
+        variable: ResultVariable[T],
+        resolver: Resolver,
+        queue_scheduler: QueueScheduler,
+    ) -> None:
         if self.resultcollector:
-            self.variable.listener(self.resultcollector, self.location)
+            variable.listener(self.resultcollector, self.location)
 
-        if self.is_ready():
-            self.target.set_value(self.target_value(), self.location)
+        if variable.is_ready():
+            self.write_target(variable)
         else:
-            requires[self] = self.variable
-            # reschedule on the variable, XU will assign it to the target variable
-            RawUnit(queue_scheduler, resolver, requires, self)
+            # reschedule on the variable's completeness
+            resumer: RawResumer = VariableReadResumer(self)
+            self.copy_location(resumer)
+            RawUnit(queue_scheduler, resolver, {self: variable}, resumer)
 
-    def execute(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler) -> T:
-        assert self.is_ready()
-        assert self.variable
-        # Attribute is ready, return it,
-        return self.variable.get_value()
 
-    def __str__(self) -> str:
-        if not self.instance:
-            return self.attribute
-        return "%s.%s" % (self.instance, self.attribute)
+class VariableReadResumer(RawResumer):
+    """
+    Resumes execution when the variable is complete.
+    """
+
+    def __init__(self, reader: VariableReader) -> None:
+        self.reader: VariableReader = reader
 
-    def get_location(self) -> Location:
-        return self.location
+    def resume(self, requires: Dict[object, ResultVariable], resolver: Resolver, queue_scheduler: QueueScheduler) -> None:
+        self.reader.write_target(requires[self.reader])
 
+    # TODO: does this ever get called? Can we get rid of it?
+    def execute(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler) -> T:
```

Comment:
The execute (and in fact the whole resume mechanism) exists to avoid the split between `VariableReader`  and `VariableReadResumer` .

`resume` is merely a second stage execute, avoiding instantiation.

Reply:
I need to dig into this a bit more, it still feels strange to me and I'm not sure I've handled this correctly at the moment.